### PR TITLE
[SPARK-23353][CORE] Allow ExecutorMetricsUpdate events to be logged t…

### PR DESCRIPTION
**Pull request by LantaoJin**
…o the event log with sampling

## What changes were proposed in this pull request?
[SPARK-22050|https://issues.apache.org/jira/browse/SPARK-22050] give a way to log BlockUpdated events. Actually, the ExecutorMetricsUpdates are also very useful if it can be persisted for further analysis. 
As a performance reason and actual use case, the PR offers a fraction configuration which can sample the events to be persisted. we also refactor for BlockUpdated with the same sampling way.

## How was this patch tested?
Unit tests

### CODE REVIEW

**Comment by LantaoJin** [comment]
@jerryshao Could you have a time to help to review?

**Review by jerryshao — COMMENTED** [review]
The patch itself looks OK to me (BTW, you need to add configurations to doc). But I don't have a strong preference from feature wise. Maybe others could have better thoughts on it.


CC @squito @vanzin .

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -53,10 +53,21 @@ package object config {
...
-  private[spark] val EVENT_LOG_BLOCK_UPDATES =
-    ConfigBuilder("spark.eventLog.logBlockUpdates.enabled")
-      .booleanConf
-      .createWithDefault(false)
+  private[spark] val EVENT_LOG_BLOCK_UPDATES_FRACTION =
+    ConfigBuilder("spark.eventLog.logBlockUpdates.fraction")
+      .doc("Expected number of times each blockUpdated event is chosen to log, " +
+        "fraction must be [0, 1]. 0 by default, means disabled")
+      .doubleConf
+      .checkValue(_ >= 0, "The fraction must not be negative")
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Ok, actually greater than 1 has the same meaning with equals to 1 (100% selected)

File: core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala [hunk 2]
```diff
@@ -228,14 +231,23 @@ private[spark] class EventLoggingListener(
     logEvent(event, flushLogger = true)
   }
 
+  // Only sampled logging when fraction greater than zero, fraction equals 1 means logging all
   override def onBlockUpdated(event: SparkListenerBlockUpdated): Unit = {
-    if (shouldLogBlockUpdates) {
-      logEvent(event, flushLogger = true)
+    if (blockUpdateSampleFraction > 0) {
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: sure

[the remark this answers is not in the document]
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala]: Think it again. Seems check `>0` could avoid generate a random number when it is configured to 0.

**Comment by jiangxb1987** [comment]
Emmm... in case we want to sample more events, does that means we shall add a new config for each event sampling?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: It is actually hard for users to set a sample ratio to balance the event log size and analysis requirement, how about control the max number of events recorded per time split?

**Comment by jiangxb1987** [comment]
I'm also worried that if we want to sample more events in the future, we have to add more configs following this way, which doesn't sound like a perfect choice.

**Comment by jerryshao** [comment]
I agree with @jiangxb1987 . @LantaoJin would you please elaborate the usage scenario of dumping executor metrics to event log? Seems history server doesn't leverage such information necessarily.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Agree, I think a max limitation is necessary.

**Comment by LantaoJin** [comment]
@jerryshao and @jiangxb1987 , thanks for your advice. In 2.1.x, the two Update events (BlockUpdated & ExecutorMetricsUpdate) are all dumb. And in 2.2.x, only BlockUpdated event has a configuration to be logged.  So the fair way is just adding an enable configuration to logging ExecutorMetrics for further using. But actually, we are refactoring the heartbeat to report more Executor information to the Driver. And this information will be logged to event log to be analysed. You know the Update events are so mass and they are not sequential. So we don't need all the events to do analysing. For example, we report Heap Usage after GC of all Executors, 10% events also can help us to get the result we want. So just add a configuration "enabled" for ExecutorMetricsUpdate is poorly simple in a big production cluster. BTW, the work I mentioned about will be contributed back to community as well soon.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: >how about control the max number of events recorded per time split?

I think this approach is still hard to balance the user requirement and event log size. Spark will possibly ignore the events that is required by the user at the specific time.

IMO, using "true" or "false" might be a feasible solution - whether to dump all the events or just ignore them. For normal user, by default (false) should be enough for them, but if you want further analysis, you can enable this by taking the risk of large event file.

For the configuration, I think we could use something like "spark.eventLog.logVerboseEvent.enabled" to control all the verbose events that will be dumped manually.

[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: "spark.eventLog.logVerboseEvent.enabled" sounds good. But this is hard to work in production environment. The risk is the evil in production environment. Sample with fraction 1 also equals to persist all events.

[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: It is not only for debugging, but also for deploying to all applications. So we need a safe way to do that. That's max limitation seems much acceptable than just enabled or disabled.

[the remark this answers is not in the document]
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: But if you're using sampling, you will possibly miss some events, for example if you're tracking memory usage, you may miss the peak memory usage events that might be important to your analysis.

Ideally, I think you can write a custom SparkListener to dump executor metrics to some time series DB like openTSDB, which might be better compared to analysis the static files.

[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Yes. I know a custom SparkListener can do the dump. But actually we can dump to eventlog file. So the conclusion is that sample is unacceptable? Is adding a configuration like "spark.eventLog.logVerboseEvent.enabled" to enable it enough?

[the remark this answers is not in the document]
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: I'm inclined to "spark.eventLog.logVerboseEvent.enabled". But I don't have a strong preference, maybe other reviewers have different thoughts about it.

[the remark this answers is not in the document]
Comment (LantaoJin) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Thank you sir. Will wait more comments.

**Comment by squito** [comment]
I can see why you want this sometimes, but I'm trying to figure out if its really valuable for users in general.  You could always add a custom listener to log this info.  It would go into separate file, not the std event log file, which means you'd have a little more work to do to stitch them together.  OTOH that could be a good thing, as it means these history server wouldn't have to parse those extra lines.

**Comment by LantaoJin** [comment]
Thanks everyone. So just close it? Or easily leave an enabled switch like blockUpdated dose? I am all OK.

**Comment by jerryshao** [comment]
I would suggest to do it like what we have already done for block update event. Since we already opened a door for block update event, it is also acceptable to leave room for another event. User should be responsible for the big event log file.

@jiangxb1987 what is your opinion?

**Comment by LantaoJin** [comment]
Thanks @jerryshao. Changed to simply add a new configuration instead sampling logging.

**Comment by jiangxb1987** [comment]
I'm still wondering whether event log is supposed to work this way, that as the source for customized analysis. I really feel we shall need some event/metrics logging framework that serves for external analysis, but we really need investigate into more use cases and give a more thoughtful design on that issue.

**Comment by squito** [comment]
I agree with @jiangxb1987 ... we already have issues with event logs being too big, as it the driver gets backlogged even writing them out, and then the history server takes a long time to parse those files.  There have been recent improvements to that, but doesn't mean we should reintroduce the problem.

I'm not saying this doesn't have a use, I'd just like to figure out if this the best way to do it.  If it only has one very specific use case for @LantaoJin , then maybe they have an alternative still using public apis, with a custom listener as I suggested.  I worry a user might turn this on (why not, more data is better) and then later on hit other scaling challenges and not realize this was the problem.

Or if this does have some general use case for all users, then maybe its fine, but I haven't seen that yet.  And maybe there is a better way to do that ... do we need another way to get detailed output metrics from executors, that doesn't have some of the scaling challenges of the event log?

**Pull request closed without merging**